### PR TITLE
Show proper metadata including album and artist whenever possible

### LINKED ISSUE #209: artist, album, song title metadata
When I download a song, it would be sweet if the metadata would already be filled out... Is that something thats not too difficult?

### PR DESCRIPTION
**Pull request by vn-ki**
Now fixes metadata when playing songs. This could be extended to fixing metadata (track title, album, artist, album art) when downloading songs. But that would add mutagen as a dependency; so I was not sure of implementing that (That could be a future PR). Switches to old method of displaying thumbnails if no proper metadata was found. Please see attached image. Fixes #209 and #495 .

image

### CODE REVIEW

**Comment by ids1024** [comment]
> This could be extended to fixing metadata (track title, album, artist, album art) when downloading songs. But that would add mutagen as a dependency

Currently `mps_youtube/commands/download.py` has a very primitive parser (`extract_metadata`) to try to get artist and title from songs, and uses avconv/ffmpeg to add the metadata.

But of course, this solution with lastfm is better, and sharing the metadata through mpris is great too.

File: mps_youtube/player.py [hunk 1]
```diff
@@ -374,14 +375,24 @@ def _get_input_file():
 def _launch_player(song, songdata, cmd):
     """ Launch player application. """
 
+    #song.title
```
Comment (hrnr) [inline mps_youtube/player.py]: some debug?

File: mps_youtube/player.py [hunk 2]
```diff
@@ -374,14 +375,24 @@ def _get_input_file():
...
-    arturl = "https://i.ytimg.com/vi/%s/default.jpg" % song.ytid
+    metadata = _get_metadata(song.title)
+
+    if metadata == None :
+        arturl = "https://i.ytimg.com/vi/%s/default.jpg" % song.ytid
+        metadata = (song.ytid, song.title, song.length, arturl, [''], '')
+    else :
+        arturl = metadata['album_art_url']
+        temp = (song.ytid, metadata['track_title'], song.length, arturl, [metadata['artist']], metadata['album'])
+        metadata = temp
```
Comment (hrnr) [inline mps_youtube/player.py]: we can assign to metadata directly, right?

**Comment by hrnr** [comment]
Thanks again for this. Seems to work reasonably well for most of the songs. Is usage of last.fm API anyhow limited? I'm just worried if all of our users start to use this, we might get blocked.

For this reason, there should be a way to change the API key. It can be similar to what we have for youtube API key.

**Comment by vn-ki** [comment]
@hrnr I don't think the last.fm API would be problem. Their [terms of service](https://www.last.fm/api/tos) talk nothing about the quantity of requests made. But we will have to give them proper credits for their data. I was thinking about implementing this for downloads too. Is dependency on mutagen a problem?

File: mps_youtube/player.py (hunk 1 above)
Comment (vn-ki) [reply mps_youtube/player.py]: Ya! Sorry.

File: mps_youtube/player.py (hunk 2 above)
Comment (vn-ki) [reply mps_youtube/player.py]: I don't know why I did that. XD

**Comment by hrnr** [comment]
Thanks. I have tested the MPRIS side and all looks good now (all datatypes correct etc.).

**Comment by ids1024** [comment]
> Is dependency on mutagen a problem?

I'm hesitant to add non-optional dependencies, but mutagen seems to be commonly packaged by distributions (Arch and Debian, at least), so it should be fine.

**Comment by Razesdark** [comment]
We could also implement the musicbrainz project as a source for music information instead of last.fm. 
Musicbrainz is an open source project and the data is released under CC0, without any need to mention the musicbrainz project when using the data. 

It also does not require an API key so the functionality won't break if the hardcoded key is revoked in the future. We're also not attributing them when using their data(we also need to show a link to https://last.fm).

**Comment by vn-ki** [comment]
@Razesdark Do they provide the album art with the json response? Their web service don't expose many things. Is there a way to search using both title and artist? I tried a few songs and the top result don't match (Which means more work to find the correct result). If there is way to search both title and artist this would improve. But still I couldn't see the ablum art in the json response. I have seen many open source projects using last.fm (Jockey and Phonograph, to name some).

**Pull request merged**